File: app/src/main/java/com/rssclient/controllers/RssItemsAdapter.java
package com.rssclient.controllers;

import java.util.ArrayList;

import com.example.alexeyglushkov.streamlib.progress.ProgressListener;
import com.aglushkov.taskmanager_http.image.Image;
import com.example.alexeyglushkov.streamlib.progress.ProgressInfo;
import com.example.alexeyglushkov.tools.HandlerTools;
import com.rssclient.model.RssItem;

import android.content.Context;

import androidx.core.content.ContextCompat;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

public class RssItemsAdapter extends ArrayAdapter<RssItem> {
    private final Context context;
    protected final ArrayList<RssItem> values;
    private RssItemsAdapterListener listener;

    class ViewHolder {
        public TextView text;
        public ImageView imageView;
        public Image loadingImage;
        public int position;
        public ProgressBar progressBar;
    }

    public RssItemsAdapter(Context context, ArrayList<RssItem> values) {
        super(context, R.layout.feed_cell, values);
        this.context = context;
        this.values = values;
    }

    public RssItemsAdapterListener getListener() {
        return listener;
    }

    public void setListener(RssItemsAdapterListener listener) {
        this.listener = listener;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {

        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            View rowView = inflater.inflate(R.layout.feed_cell, parent, false);
            TextView textView = (TextView) rowView.findViewById(R.id.topLine);
            ImageView imageView = (ImageView) rowView.findViewById(R.id.icon);

            ViewHolder holder = new ViewHolder();
            holder.text = textView;
            holder.imageView = imageView;

            holder.progressBar = (ProgressBar)rowView.findViewById(R.id.progress);
            holder.progressBar.setIndeterminate(false);

            convertView = rowView;
            convertView.setTag(holder);
        }

        final ViewHolder holder = (ViewHolder) convertView.getTag();

        RssItem item = values.get(position);
        holder.text.setText(position + ". " + item.title());
        holder.position = position;

        if (item.image() != null) {
            holder.progressBar.setVisibility(View.VISIBLE);
            loadImage(convertView, item);
        } else {
            holder.progressBar.setVisibility(View.INVISIBLE);
            holder.imageView.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_launcher));
        }

        return convertView;
    }

    protected void loadImage(View convertView, RssItem item) {
        final ViewHolder holder = (ViewHolder) convertView.getTag();

        holder.imageView.setImageDrawable(null);
        holder.loadingImage = item.image();
        if (item.image() == null) {
            return;
        }

        final int position = holder.position;
        holder.progressBar.setProgress(0);

        getListener().loadImage(item);
    }

    public interface RssItemsAdapterListener {
        void loadImage(RssItem item);
    }
}

File: app/src/main/java/com/rssclient/controllers/RssItemsAdapter.kt
package com.rssclient.controllers

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.aglushkov.taskmanager_http.image.Image
import com.rssclient.model.RssItem

class RssItemsAdapter(context: Context, val values: ArrayList<RssItem>) : ArrayAdapter<RssItem>(context, R.layout.feed_cell, values) {
    var listener: RssItemsAdapterListener? = null

    internal inner class ViewHolder {
        lateinit var text: TextView
        lateinit var imageView: ImageView
        var loadingImage: Image? = null
        lateinit var progressBar: ProgressBar
        var position = 0
    }

    override fun getView(position: Int, convertView: View, parent: ViewGroup): View {
        var convertView = convertView
        if (convertView == null) {
            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val rowView = inflater.inflate(R.layout.feed_cell, parent, false)
            val textView = rowView.findViewById<View>(R.id.topLine) as TextView
            val imageView = rowView.findViewById<View>(R.id.icon) as ImageView

            val holder = ViewHolder()
            holder.text = textView
            holder.imageView = imageView
            holder.progressBar = rowView.findViewById<View>(R.id.progress) as ProgressBar
            holder.progressBar.isIndeterminate = false
            convertView = rowView
            convertView.tag = holder
        }
        val holder = convertView.tag as ViewHolder
        val item = values[position]
        holder.text.text = position.toString() + ". " + item.title
        holder.position = position
        if (item.image != null) {
            holder.progressBar.visibility = View.VISIBLE
            loadImage(convertView, item)
        } else {
            holder.progressBar.visibility = View.INVISIBLE
            holder.imageView.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_launcher))
        }
        return convertView
    }

    protected fun loadImage(convertView: View, item: RssItem) {
        val holder = convertView.tag as ViewHolder
        holder.imageView.setImageDrawable(null)
        holder.loadingImage = item.image
        if (item.image == null) {
            return
        }
        val position = holder.position
        holder.progressBar.progress = 0
        listener?.loadImage(item)
    }

    interface RssItemsAdapterListener {
        fun loadImage(item: RssItem)
    }

}